src/hotspot/share/gc/shenandoah/shenandoahMetrics.cpp

```diff
@@ -46,8 +46,10 @@ void ShenandoahMetricsSnapshot::snap_after() {
 
 bool ShenandoahMetricsSnapshot::is_good_progress() {
   // Under the critical threshold?
-  size_t free_actual   = _heap->free_set()->available();
-  size_t free_expected = _heap->max_capacity() / 100 * ShenandoahCriticalFreeThreshold;
+  ShenandoahFreeSet* free_set = _heap->free_set();
+  size_t free_actual   = free_set->available();
+  // The sum of free_set->capacity() and ->reserved represents capacity of young in generational, heap in non-generational.
+  size_t free_expected = ((free_set->capacity() + free_set->reserved()) / 100) * ShenandoahCriticalFreeThreshold;
```

Comment:
As an outsider, the units involved and what exactly is being calculated is pretty opaque. Why would we divide by 100 to compute free_expected and not do the same for free_actual? Do we care about integer division truncation? The default value of ShenandoahCriticalFreeThreshold is 1, so multiplying by it is a nop by default, which seems strange.

Reply:
ShenandoahCriticalFreeThreshold represents a percentage of the "total size".  To calculate N% of the young generation size, we divide the generation size by 100 and then multiply by ShenandoahCriticalFreeThreshold.  This code is a bit different in the most recent revision.  Do you think it needs a comment?

Reply:
Yes :)